packages/rabbitmq/src/amqp/connection.ts

```diff
@@ -376,11 +363,75 @@ export class AmqpConnection {
       throw new Error('AMQP connection is not available');
     }
 
-    this._channel.publish(
-      exchange,
-      routingKey,
-      Buffer.from(JSON.stringify(message)),
-      options
-    );
+    let buffer: Buffer;
+    if (message instanceof Buffer) {
+      buffer = message;
+    } else if (message instanceof Uint8Array) {
+      buffer = Buffer.from(message);
+    } else if (message != null) {
+      buffer = Buffer.from(JSON.stringify(message));
+    } else {
+      buffer = Buffer.alloc(0);
+    }
+
+    this._channel.publish(exchange, routingKey, buffer, options);
+  }
+
+  private handleMessage<T, U>(
+    handler: (
+      msg: T | undefined,
+      rawMessage?: amqplib.ConsumeMessage
+    ) => Promise<U>,
+    msg: amqplib.ConsumeMessage,
+    allowNonJsonMessages?: boolean
+  ) {
+    let message: T | undefined = undefined;
+    if (msg.content) {
+      if (allowNonJsonMessages) {
+        try {
+          message = JSON.parse(msg.content.toString()) as T;
+        } catch {
+          // Let handler handle parsing error, it has the raw message anyway
+          message = undefined;
+        }
+      } else {
+        message = JSON.parse(msg.content.toString()) as T;
+      }
+    }
+
+    return handler(message, msg);
+  }
+
+  private async handleError(
+    channel: amqplib.Channel,
+    msgOptions: MessageHandlerOptions,
+    errorBehavior: MessageHandlerErrorBehavior,
+    msg: amqplib.Message,
+    error: any
+  ) {
+    if (msg == null) {
+      return;
+    } else {
+      try {
+        if (msgOptions.errorCallbacks) {
+          for (const callback of msgOptions.errorCallbacks) {
+            await callback(channel, msg, error);
+          }
+        }
+      } finally {
```

Comment:
I feel like at this point we should think about removing this finally case and replacing the switch case with new named Error Handling Functions that represent the old behaviors, similar to the ReplyErrorCallback that you made. If we run the `msgOptions.errorCallbacks` as well as one from the legacy behaviors here we we will likely try to ack or nack a message that has already been acked or nacked which could lead to more problems

Reply:
Should we catch errors in handlers and continue other handler execution if one fails?